### PR TITLE
FIX: completeMPU

### PR DESCRIPTION
**Pull request by nicolas2bert**
FIXES(#789)

### CODE REVIEW

File: lib/api/completeMultipartUpload.js [hunk 1]
```diff
@@ -258,107 +259,121 @@ function completeMultipartUpload(authInfo, request, log, callback) {
...
-                // return an error
+                // in ascending order return an error
                 const partNumber = Number.parseInt(part.PartNumber[0], 10);
-                if (partNumber !== i + 1) {
-                    // If not in order return InvalidPartOrder
-                    if (partNumber <= partLength) {
+                if (i > 0) {
+                    const previousPartNumber =
+                      Number.parseInt(jsonList.Part[i - 1].PartNumber[0], 10);
+                    if (partNumber <= previousPartNumber) {
```
Comment (LaurenSpiegel) [inline lib/api/completeMultipartUpload.js]: If you send a complete mpu call with an inventory that states a part number more than once, what error does aws send?

Comment (mikeys) [inline lib/api/completeMultipartUpload.js]: @nicolas2bert I'm not 100% sure that there's a restriction on the parts order in the MPU Complete request.

E.g. 
```json
{
  "Part": [
    { "PartNumber": 8, "ETag": "..." },
    { "PartNumber": 5, "ETag": "..." },
    { "PartNumber": 13, "ETag": "..." }
  ]
}
```

May be legit. All that I could grasp from the AWS Doc is that the items are concatenated based on the part numbers.

Comment (nicolas2bert) [inline lib/api/completeMultipartUpload.js]: Hi @mikeys, 

There is an actual meaning to the array order.

Tested against AWS:

```
{
  "Part": [
    { "PartNumber": 1, "ETag": "..." },
    { "PartNumber": 2, "ETag": "..." },
    { "PartNumber": 3, "ETag": "..." }
  ]
}
```
returns no error.
However, 
```
{
  "Part": [
    { "PartNumber": 1, "ETag": "..." },
    { "PartNumber": 3, "ETag": "..." },
    { "PartNumber": 2, "ETag": "..." }
  ]
}
```
returns 400 - InvalidPartOrder error

Comment (mikeys) [inline lib/api/completeMultipartUpload.js]: @nicolas2bert You've checked it against the actual AWS S3 Service? cool. Thanks for getting into this.

Comment (nicolas2bert) [inline lib/api/completeMultipartUpload.js]: yes I did @mikeys, sure 👍

File: lib/api/completeMultipartUpload.js [hunk 2]
```diff
@@ -258,107 +266,127 @@ function completeMultipartUpload(authInfo, request, log, callback) {
...
-                // the ETag for the part already sent, return an error
-                if (partETag !== storedPartsAsObjects[i].ETag) {
-                    return next(errors.InvalidPart, destBucket);
-                }
-                // If any part other than the last part is less than 5MB,
-                // return an error
-                const partSize = Number.parseInt(storedPartsAsObjects[i]
-                    .size, 10);
+                let isPartUploaded = false;
+                while (storedPartsAsObjects.length > 0 && !isPartUploaded) {
```
Comment (electrachong) [inline lib/api/completeMultipartUpload.js]: if you prefer, you can do:
```
let isPartUploaded = false
for ( ; storedPartsAsObjects.length > 0 && !isPartUploaded; storedPartsAsObjects = storedPartsAsObjects.splice(1));
```
Go with what you think is most readable!

File: lib/api/completeMultipartUpload.js [hunk 3]
```diff
@@ -221,18 +222,25 @@ function completeMultipartUpload(authInfo, request, log, callback) {
...
             }
-            const storedPartsAsObjects = storedParts.map(item => ({
-                // In order to delete the part listing in the shadow
-                // bucket, need the full key
-                key: item.key,
-                ETag: `"${item.value.ETag}"`,
-                size: item.value.Size,
+            let storedPartsAsObjects = [];
+            const keysToDelete = [];
+            for (let i = 0; i < storedParts.length; i++) {
```
Comment (nicolas2bert) [inline lib/api/completeMultipartUpload.js]: Yes forEach is probably more appropriate.
Interesting article I found: http://thejsguy.com/2016/07/30/javascript-for-loop-vs-array-foreach.html

File: lib/api/completeMultipartUpload.js (hunk 2 above)
Comment (nicolas2bert) [inline lib/api/completeMultipartUpload.js]: I think it is readable enough. If you don't think so, I can change it.

File: lib/api/completeMultipartUpload.js [hunk 4]
```diff
@@ -258,107 +266,127 @@ function completeMultipartUpload(authInfo, request, log, callback) {
...
+                                cipheredDataKey: location.sseCipheredDataKey,
+                            };
+                            dataLocations.push(pieceRetrievalInfo);
+                            calculatedSize += pieceSize;
+                        }
+
+                        const partETagWithoutQuotes =
+                            storedPart.ETag.slice(1, -1);
+                        concatETags += partETagWithoutQuotes;
+                        // partUploadedCompleted.push(storedPart);
```
Comment (nicolas2bert) [inline lib/api/completeMultipartUpload.js]: nope, thanks

**Comment by nicolas2bert** [comment]
@LaurenSpiegel @electrachong I adressed your comments here :)

**Pull request merged**